### DISCUSSION TITLE
New package zustand-scoped that tackles nested stores and initializing stores with data

### DISCUSSION THREAD

**Discussion by lukesmurray** [pmndrs/zustand#1587] [Show and tell]
I wanted to publicize a package I created, ["zustand-scoped"](https://github.com/lukesmurray/zustand-scoped) which attempts to tackle two issues I've had with Zustand—combining stores (#291, #161, #163, #178) and initializing stores with data (#82, #552).

The package lets you

- create stores with initial data (`create(initialData)`)
- write child stores that can be used by themselves or within a parent store
- split a monolith store into several child stores split across multiple files

The package is typesafe and works with almost all zustand middleware ([not immer](https://github.com/lukesmurray/zustand-scoped#caveats)).

## How it Works

The package exposes a wrapper around `create`, which makes the return value of `create` a "store factory" instead of a hook. You can call the "factory" with initial data to create a store and set the initial data of the store.

```tsx
import { createScopedStore } from "@lukesmurray/zustand-scoped";

type BearState = {
  bears: number;
  increasePopulation: () => void;
  removeAllBears: () => void;
};

type BearInitialData = {
  bears: number;
};

// define a store factory
const createBearStore = createScopedStore<BearState, BearInitialData>()(
  (initialData) =>
    (set, get) => ({
      bears: initialData.bears,
      increasePopulation: () =>
        set((state) => ({ ...state, bears: state.bears + 1 })),
      removeAllBears: () => set({ ...get(), bears: 0 }),
    })
);

// create a hook by passing initial data.
const useBearStore = createBearStore({ bears: 0 });
```

The package also provides typesafe support for "nested" stores, inspired by  [this comment](https://github.com/pmndrs/zustand/issues/163#issuecomment-678821969) (see [scoped store docs](https://github.com/lukesmurray/zustand-scoped#scoped-stores) for details).

Compared with [zustand-lens](https://github.com/dhmk083/dhmk-zustand-lens) this is much less magical. You need to write your own `set` and `get` overrides in the parent store. But as a result "zustand-scoped" supports nesting stores with arbitrary complexity. For instance, you can create an array of nested stores and dynamically add and remove them at runtime.

The entire package is also **406B** so feel free to add it to your bundle.

Would appreciate any feedback!

**Comment by baranyildirim** [comment]
Nice! I've been looking for a way to nest-stores in a reasonable fashion and will try this out. It would be great if a solution like this was available in zustand itself. Thanks for publishing your package!
